File: src/pack_manager.py
"""
MoltCare Pack Manager
智能包管理器 - 支持安装、卸载、列出pack
"""

import json
import os
import shutil
from pathlib import Path
from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass, asdict
from datetime import datetime
import hashlib


@dataclass
class PackManifest:
    """Pack 清单文件结构"""
    name: str
    version: str
    description: str = ""
    author: str = ""
    dependencies: List[str] = None
    entry_point: str = "main.py"
    
    def __post_init__(self):
        if self.dependencies is None:
            self.dependencies = []
    
    def to_dict(self) -> Dict:
        return asdict(self)
    
    @classmethod
    def from_dict(cls, data: Dict) -> "PackManifest":
        return cls(**{k: v for k, v in data.items() if k in cls.__dataclass_fields__})


@dataclass
class PackInfo:
    """已安装Pack的信息"""
    name: str
    version: str
    install_date: str
    manifest: PackManifest
    path: str
    active: bool = True


class PackManager:
    """
    Pack 管理器 - 核心功能
    - 安装: 从本地路径或GitHub安装
    - 卸载: 安全移除pack
    - 列出: 显示已安装packs
    - 启用/禁用: 控制pack状态
    """
    
    def __init__(self, packs_dir: str = "./packs", config_path: Optional[str] = None):
        self._packs_dir = Path(packs_dir)
        self._packs_dir.mkdir(parents=True, exist_ok=True)
        self._index_path = self._packs_dir / ".index.json"
        self._installed: Dict[str, PackInfo] = {}
        self._load_index()
    
    def _load_index(self) -> None:
        """加载pack索引"""
        if self._index_path.exists():
            try:
                with open(self._index_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                for name, info_data in data.get("packs", {}).items():
                    manifest = PackManifest.from_dict(info_data.get("manifest", {}))
                    self._installed[name] = PackInfo(
                        name=name,
                        version=info_data.get("version", ""),
                        install_date=info_data.get("install_date", ""),
                        manifest=manifest,
                        path=info_data.get("path", ""),
                        active=info_data.get("active", True)
                    )
            except (json.JSONDecodeError, IOError) as e:
                print(f"[PackManager] 加载索引失败: {e}")
    
    def _save_index(self) -> bool:
        """保存pack索引"""
        try:
            data = {
                "updated_at": datetime.now().isoformat(),
                "packs": {}
            }
            for name, info in self._installed.items():
                data["packs"][name] = {
                    "version": info.version,
                    "install_date": info.install_date,
                    "manifest": info.manifest.to_dict(),
                    "path": info.path,
                    "active": info.active
                }
            with open(self._index_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            return True
        except IOError as e:
            print(f"[PackManager] 保存索引失败: {e}")
            return False
    
    def _sanitize_pack_name(self, name: str) -> Tuple[bool, str]:
        """
        净化pack名称，防止路径遍历攻击
        
        禁止的字符/模式:
        - 路径分隔符: /, \
        - 父目录引用: .. 
        - 空字符串或纯空白
        - 以点开头的隐藏文件
        - 绝对路径
        
        Returns:
            (是否有效, 净化后的名称或错误信息)
        """
        if not name or not name.strip():
            return False, "Pack名称不能为空"
        
        # 检查路径分隔符
        if '/' in name or '\\' in name:
            return False, "Pack名称不能包含路径分隔符"
        
        # 检查父目录引用
        if '..' in name:
            return False, "Pack名称不能包含'..'"
        
        # 检查隐藏文件 (以.开头)
        if name.startswith('.'):
            return False, "Pack名称不能以'.'开头"
        
        # 检查控制字符
        for char in name:
            if ord(char) < 32 or ord(char) == 127:
                return False, "Pack名称包含非法控制字符"
        
        # 净化：去除首尾空白
        sanitized = name.strip()
        
        # 检查长度限制
        if len(sanitized) > 100:
            return False, "Pack名称长度不能超过100字符"
        
        return True, sanitized

    def sanitizePackName(self, name: str) -> dict:
        """
        公共API：净化pack名称
        
        Returns:
            {
                'valid': bool,
                'name': str (净化后的名称，如果有效),
                'error': str (错误信息，如果无效)
            }
        """
        is_valid, result = self._sanitize_pack_name(name)
        if is_valid:
            return {'valid': True, 'name': result, 'error': None}
        else:
            return {'valid': False, 'name': None, 'error': result}

    def _validate_manifest(self, manifest_path: Path) -> Tuple[bool, Optional[PackManifest]]:
        """验证pack清单文件"""
        if not manifest_path.exists():
            return False, None
        try:
            with open(manifest_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            manifest = PackManifest.from_dict(data)
            if not manifest.name or not manifest.version:
                return False, None
            
            # 验证pack名称安全性 (SEC-001 修复)
            is_valid_name, result = self._sanitize_pack_name(manifest.name)
            if not is_valid_name:
                print(f"[PackManager] Pack名称验证失败: {result}")
                return False, None
            
            # 更新净化后的名称
            manifest.name = result
            
            return True, manifest
        except (json.JSONDecodeError, KeyError) as e:
            print(f"[PackManager] 清单验证失败: {e}")
            return False, None
    
    def install(self, source: str, force: bool = False) -> Tuple[bool, str]:
        """
        安装Pack
        
        Args:
            source: Pack源路径(本地目录或GitHub repo)
            force: 强制重新安装
        
        Returns:
            (成功状态, 消息)
        """
        source_path = Path(source)
        
        # 检查源是否存在
        if not source_path.exists():
            return False, f"Pack源不存在: {source}"
        
        # 查找并验证manifest
        manifest_path = source_path / "manifest.json"
        is_valid, manifest = self._validate_manifest(manifest_path)
        
        if not is_valid or manifest is None:
            return False, "无效的Pack: 缺少或损坏的manifest.json"
        
        pack_name = manifest.name
        
        # 检查是否已安装
        if pack_name in self._installed:
            if not force:
                return False, f"Pack '{pack_name}' 已安装，使用 --force 强制重装"
            # 先卸载旧版本
            self.uninstall(pack_name)
        
        # 安装目录
        install_path = self._packs_dir / pack_name
        
        try:
            # 复制文件
            if install_path.exists():
                shutil.rmtree(install_path)
            shutil.copytree(source_path, install_path, ignore=shutil.ignore_patterns('.git', '__pycache__', '*.pyc'))
            
            # 记录到索引
            self._installed[pack_name] = PackInfo(
                name=pack_name,
                version=manifest.version,
                install_date=datetime.now().isoformat(),
                manifest=manifest,
                path=str(install_path),
                active=True
            )
            
            self._save_index()
            return True, f"✓ Pack '{pack_name}' v{manifest.version} 安装成功"
            
        except Exception as e:
            # 清理失败的安装
            if install_path.exists():
                shutil.rmtree(install_path)
            return False, f"安装失败: {e}"
    
    def uninstall(self, pack_name: str) -> Tuple[bool, str]:
        """卸载Pack"""
        if pack_name not in self._installed:
            return False, f"Pack '{pack_name}' 未安装"
        
        pack_info = self._installed[pack_name]
        install_path = Path(pack_info.path)
        
        try:
            # 删除文件
            if install_path.exists():
                shutil.rmtree(install_path)
            
            # 从索引移除
            del self._installed[pack_name]
            self._save_index()
            
            return True, f"✓ Pack '{pack_name}' 已卸载"
            
        except Exception as e:
            return False, f"卸载失败: {e}"
    
    def list_packs(self, show_inactive: bool = False) -> List[PackInfo]:
        """列出已安装的Packs"""
        packs = []
        for name, info in self._installed.items():
            if show_inactive or info.active:
                packs.append(info)
        return sorted(packs, key=lambda p: p.name)
    
    def get_pack(self, pack_name: str) -> Optional[PackInfo]:
        """获取指定Pack信息"""
        return self._installed.get(pack_name)
    
    def enable(self, pack_name: str) -> Tuple[bool, str]:
        """启用Pack"""
        if pack_name not in self._installed:
            return False, f"Pack '{pack_name}' 未安装"
        
        self._installed[pack_name].active = True
        self._save_index()
        return True, f"✓ Pack '{pack_name}' 已启用"
    
    def disable(self, pack_name: str) -> Tuple[bool, str]:
        """禁用Pack"""
        if pack_name not in self._installed:
            return False, f"Pack '{pack_name}' 未安装"
        
        self._installed[pack_name].active = False
        self._save_index()
        return True, f"✓ Pack '{pack_name}' 已禁用"
    
    def is_installed(self, pack_name: str) -> bool:
        """检查Pack是否已安装"""
        return pack_name in self._installed
    
    def get_active_packs(self) -> List[PackInfo]:
        """获取所有启用的Packs"""
        return [info for info in self._installed.values() if info.active]


# 全局PackManager实例
def get_pack_manager(packs_dir: str = "./packs") -> PackManager:
    """获取全局PackManager实例"""
    return PackManager(packs_dir)


# 自我审查检查点 (累计代码行数: ~280行)
# ✅ 完整的CRUD操作
# ✅ 索引持久化
# ✅ 原子安装(失败清理)
# ✅ 启用/禁用状态管理

File: src/pack_manager.ts
/**
 * Pack Manager (TypeScript)
 * Pack 扫描和管理功能
 */

import * as fs from 'fs';
import * as path from 'path';
import Handlebars from 'handlebars';

export interface PackManifest {
  name: string;
  version: string;
  title?: string;
  description?: string;
  author?: string;
  category?: string;
  priority?: number;
  isCore?: boolean;
  createdAt?: string;
  templates?: PackTemplate[];
  scripts?: {
    apply?: string;
    preApply?: string;
    postApply?: string;
  };
  dependencies?: string[];
  config?: {
    backupExisting?: boolean;
    allowOverwrite?: boolean;
    validateTarget?: boolean;
  };
}

export interface PackTemplate {
  file: string;
  target: string;
  required?: boolean;
  description?: string;
  variables?: Record<string, string>;
}

export interface PackInfo {
  name: string;
  version: string;
  title?: string;
  description?: string;
  author?: string;
  category?: string;
  isCore?: boolean;
  manifest: PackManifest;
  path: string;
  installed: boolean;
  installDate?: string;
}

export interface PackIndexEntry {
  version: string;
  installDate: string;
  manifest: PackManifest;
  path: string;
  active: boolean;
}

export interface PackIndex {
  updatedAt: string;
  packs: Record<string, PackIndexEntry>;
}

export class PackManager {
  private packsDir: string;
  private indexPath: string;
  private index: PackIndex;

  constructor(packsDir: string) {
    this.packsDir = packsDir;
    this.indexPath = path.join(packsDir, '.index.json');
    this.index = { updatedAt: new Date().toISOString(), packs: {} };
    this.loadIndex();
  }

  /**
   * 加载索引
   */
  private loadIndex(): void {
    try {
      if (fs.existsSync(this.indexPath)) {
        const content = fs.readFileSync(this.indexPath, 'utf-8');
        this.index = JSON.parse(content) as PackIndex;
      }
    } catch (error) {
      this.index = { updatedAt: new Date().toISOString(), packs: {} };
    }
  }

  /**
   * 保存索引
   */
  private saveIndex(): boolean {
    try {
      this.index.updatedAt = new Date().toISOString();
      fs.writeFileSync(this.indexPath, JSON.stringify(this.index, null, 2), 'utf-8');
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * 净化pack名称（安全检查）
   */
  sanitizePackName(name: string): { valid: boolean; name?: string; error?: string } {
    if (!name || !name.trim()) {
      return { valid: false, error: 'Pack 名称不能为空' };
    }

    // 检查路径分隔符
    if (name.includes('/') || name.includes('\\')) {
      return { valid: false, error: 'Pack 名称不能包含路径分隔符' };
    }

    // 检查父目录引用
    if (name.includes('..')) {
      return { valid: false, error: 'Pack 名称不能包含 ".."' };
    }

    // 检查隐藏文件
    if (name.startsWith('.')) {
      return { valid: false, error: 'Pack 名称不能以 "." 开头' };
    }

    // 检查控制字符
    for (const char of name) {
      if (char.charCodeAt(0) < 32 || char.charCodeAt(0) === 127) {
        return { valid: false, error: 'Pack 名称包含非法控制字符' };
      }
    }

    const sanitized = name.trim();

    if (sanitized.length > 100) {
      return { valid: false, error: 'Pack 名称长度不能超过 100 字符' };
    }

    return { valid: true, name: sanitized };
  }

  /**
   * 扫描可用的 packs
   */
  scanPacks(): PackInfo[] {
    const packs: PackInfo[] = [];

    if (!fs.existsSync(this.packsDir)) {
      return packs;
    }

    const entries = fs.readdirSync(this.packsDir, { withFileTypes: true });

    for (const entry of entries) {
      if (!entry.isDirectory() || entry.name.startsWith('.')) {
        continue;
      }

      const packPath = path.join(this.packsDir, entry.name);
      const manifestPath = path.join(packPath, 'manifest.json');

      if (!fs.existsSync(manifestPath)) {
        continue;
      }

      try {
        const content = fs.readFileSync(manifestPath, 'utf-8');
        const manifest = JSON.parse(content) as PackManifest;

        // 验证名称
        const validation = this.sanitizePackName(manifest.name || entry.name);
        if (!validation.valid) {
          continue;
        }

        const installedEntry = this.index.packs[manifest.name];

        packs.push({
          name: manifest.name || entry.name,
          version: manifest.version || '0.0.0',
          title: manifest.title,
          description: manifest.description,
          author: manifest.author,
          category: manifest.category,
          isCore: manifest.isCore,
          manifest,
          path: packPath,
          installed: !!installedEntry,
          installDate: installedEntry?.installDate,
        });
      } catch (error) {
        // 忽略损坏的 manifest
      }
    }

    // 按优先级排序（core packs 优先）
    return packs.sort((a, b) => {
      if (a.isCore && !b.isCore) return -1;
      if (!a.isCore && b.isCore) return 1;
      const priorityA = a.manifest.priority ?? 999;
      const priorityB = b.manifest.priority ?? 999;
      return priorityA - priorityB;
    });
  }

  /**
   * 获取指定 pack
   */
  getPack(name: string): PackInfo | undefined {
    return this.scanPacks().find(p => p.name === name);
  }

  /**
   * 检查 pack 是否已安装
   */
  isInstalled(name: string): boolean {
    return name in this.index.packs;
  }

  /**
   * 获取所有 pack 名称
   */
  getPackNames(): string[] {
    return this.scanPacks().map(p => p.name);
  }

  /**
   * 获取已分类的 packs
   */
  getPacksByCategory(): Record<string, PackInfo[]> {
    const packs = this.scanPacks();
    const categories: Record<string, PackInfo[]> = {};

    for (const pack of packs) {
      const category = pack.category || 'other';
      if (!categories[category]) {
        categories[category] = [];
      }
      categories[category].push(pack);
    }

    return categories;
  }

  /**
   * 渲染模板
   */
  renderTemplate(templatePath: string, variables: Record<string, unknown>): string {
    const content = fs.readFileSync(templatePath, 'utf-8');
    const template = Handlebars.compile(content);
    return template(variables);
  }

  /**
   * 安装 pack（标记为已安装）
   */
  install(name: string): boolean {
    const pack = this.getPack(name);
    if (!pack) return false;

    this.index.packs[name] = {
      version: pack.version,
      installDate: new Date().toISOString(),
      manifest: pack.manifest,
      path: pack.path,
      active: true,
    };

    return this.saveIndex();
  }

  /**
   * 获取所有已安装 pack 名称
   */
  getInstalledNames(): string[] {
    return Object.keys(this.index.packs);
  }

  /**
   * 获取索引
   */
  getIndex(): PackIndex {
    return { ...this.index };
  }
}
